### PR TITLE
[IIDM v1.3] Fix TieLineAdder (#192)

### PR DESCRIPTION
**Pull request by sebalaig**
Signed-off-by: Sébastien LAIGRE <slaigre@silicom.fr>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#192 

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

**Review by mathbagu — CHANGES_REQUESTED** [review]

File: src/iidm/TieLineAdder.cpp [hunk 1]
```diff
@@ -28,14 +28,25 @@ TieLine& TieLineAdder::add() {
...
+    }
+    if (!m_halfLineAdder2.is_initialized()) {
+        throw ValidationException(*this, "half line 2 is not set");
+    }
+
+    TieLine::HalfLine half1 = m_halfLineAdder1->build();
+    TieLine::HalfLine half2 = m_halfLineAdder2->build();
+
+    checkHalf(*this, half1, 1);
+    checkHalf(*this, half2, 2);
```
Comment (mathbagu) [inline src/iidm/TieLineAdder.cpp:43]: The checks are already done in the HalfLineAdder::add()?

File: src/iidm/TieLine.cpp [hunk 2]
```diff
@@ -16,119 +16,10 @@ namespace powsybl {
...
-    return *this;
-}
-
-TieLine::TieLine(const std::string& id, const std::string& name, bool fictitious, const std::string& ucteXnodeCode, HalfLine& half1, HalfLine& half2) :
+TieLine::TieLine(const std::string& id, const std::string& name, bool fictitious, const std::string& ucteXnodeCode, HalfLine&& half1, HalfLine&& half2) :
     Line(id, name, fictitious),
-    m_half1(attach(half1)),
-    m_half2(attach(half2)),
+    m_half1(std::move(attach(half1))),
+    m_half2(std::move(attach(half2))),
```
Comment (mathbagu) [inline src/iidm/TieLine.cpp:22]: To be consistent with the DanglingLine and ShuntCompensator, I would attach the TieLine to the half line after assign it

File: src/iidm/HalfLineAdder.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    halfLine.setFictitious(m_fictitious);
+    halfLine.setXnodeP(m_xnodeP);
+    halfLine.setXnodeQ(m_xnodeQ);
+    halfLine.setR(m_r);
+    halfLine.setX(m_x);
+    halfLine.setG1(m_g1);
+    halfLine.setG2(m_g2);
+    halfLine.setB1(m_b1);
+    halfLine.setB2(m_b2);
+    return halfLine;
```
Comment (mathbagu) [inline src/iidm/HalfLineAdder.cpp:81]: Use a constructor?!

File: include/powsybl/iidm/HalfLine.hpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    void setName(const std::string& name);
+
+    void setParent(TieLine& parent);
+
+    friend class iidm::TieLine;
+
+    friend class iidm::TieLineAdder;
+
+    friend class HalfLineAdder;
```
Comment (mathbagu) [inline include/powsybl/iidm/HalfLine.hpp:87]: This should not be useful

File: include/powsybl/iidm/HalfLine.hpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    HalfLine& setR(double r);
+
+    HalfLine& setX(double x);
+
+    HalfLine& setXnodeP(double xnodeP);
+
+    HalfLine& setXnodeQ(double xnodeQ);
+
+private:
+    HalfLine();
```
Comment (mathbagu) [inline include/powsybl/iidm/HalfLine.hpp:75]: Change the halfLine constructor to initialize all the fields

**Review by mathbagu — CHANGES_REQUESTED** [review]

File: include/powsybl/iidm/HalfLine.hpp (hunk 4 above)
Comment (sebalaig) [reply include/powsybl/iidm/HalfLine.hpp:87]: Yes it is, in order to call HalfLine constructor which is private.

File: src/iidm/TieLine.cpp (hunk 2 above)
Comment (sebalaig) [reply src/iidm/TieLine.cpp:22]: Ok I call setParent on both half Line in constructor body.

File: include/powsybl/iidm/HalfLine.hpp (hunk 4 above)
Comment (mathbagu) [reply include/powsybl/iidm/HalfLine.hpp:88]: Could you please create a PR in the v1.3.1 milestone: I think we are not consistent with the visibility of our constructors

File: src/iidm/HalfLine.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+void HalfLine::setId(const std::string& id) {
+    m_id = id;
+}
+
+void HalfLine::setName(const std::string& name) {
+    m_name = name;
+}
+
+void HalfLine::setParent(TieLine& parent) {
+    m_parent = stdcxx::ref(parent);
```
Comment (mathbagu) [inline src/iidm/HalfLine.cpp:110]: m_parent = parent

**Review by mathbagu — APPROVED** [review]

**Pull request merged**